### DISCUSSION TITLE
classes/class_touchscreenbutton

### DISCUSSION THREAD

**Discussion by giscus** [godotengine/godot-docs-user-notes#80] [User-contributed notes]
# classes/class_touchscreenbutton

Inherits: Node2D< CanvasItem< Node< Object Button for touch screen devices for gameplay use. Description: TouchScreenButton allows you to create on-screen buttons for touch devices. It's intended f...

https://docs.godotengine.org/en/stable/classes/class_touchscreenbutton.html

**Comment by lukky-dev** [comment]
"Unlike with Control nodes, you cannot set anchors on it."

In case you do want to anchor the touchscreenbutton to anything other then the top left corner of the screen you can make the node a child of a Control node that has it's anchors set.

I spend some time scratching my head on how to make right handed touchscreenbuttons responsive, hope this saves you some time!

**Reply by joseolinda** [reply]
Thanks! You help a lot!

**Reply by Keyvan-Rostyslav** [reply]
Hello, Can you send a screenshot of how you did it?

**Reply by Calinou** [reply]
This is what they mean:

![image](https://github.com/user-attachments/assets/c93e971c-d413-4c3a-b86e-18d17bf4e58b)

Note that you must set the *TouchscreenContainer* nodes' custom minimum size to match the texture size of their TouchScreenButton children. Otherwise, anchors and containers won't work as expected and the buttons will overlap.

**Reply by JEnriquezG88** [reply]
You're a hero.

**Comment by JimmyCushnie** [comment]
"Unlike with Control nodes, you cannot set anchors on it."

Besides the solution shared by @lukky-dev, you can also intercept the touch events directly. Here's a c# example:

```csharp
using Godot;

public partial class TouchButton : Control
{
    public override void _Input(InputEvent @event)
    {
        switch (@event)
        {
            case InputEventScreenTouch touchEvent:
                if (GetGlobalRect().HasPoint(touchEvent.Position))
                {
                    if (touchEvent.Pressed)
                        GD.Print($"touched down!");
                    else
                        GD.Print($"touched up!");

                    GetViewport().SetInputAsHandled();
                }
                break;

            case InputEventScreenDrag dragEvent:
                if (GetGlobalRect().HasPoint(dragEvent.Position))
                {
                    GD.Print($"dragged!");
                    GetViewport().SetInputAsHandled();
                }
                break;
        }
    }
}
```

**Comment by andOlga** [comment]
> Actions can be handled with InputEventAction.

They don't *have* to be, though. Handling them via the `Input` singleton works fine too (with functions like `Input.is_action_just_pressed()`).

**Comment by superknight845-creator** [comment]
What does mean "action" in its property's?

**Reply by Calinou** [reply]
The `action` property is the name of the input action that is emitted when the button is pressed. For example, if you have an action called `jump` in your Project Settings' input map tab, you should set the `action` property to `jump`.

**Comment by dhAlcojor** [comment]
Hi! I have added this button in my game to trigger the "shoot" Input Action. For some reason, when I push this button in Android, the player shoots twice. This doesn't happen if I keep the button visible while debugging and I click on it.

I guess this might be because the action is being triggered both in the "pressed" and "released" signals, but I don't know how to prevent it since I'm not handling the triggering myself (it's being done via the "Action" property in the Inspector).

I have tried hijacking the "released" signal:

```gdscript
_mobile_fire_button.released.connect(func():
    get_tree().set_input_as_handled()
)
```

But it didn't work.
